src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -53,40 +53,61 @@ public Assembly LoadFromPath(string fullPath)
 
         private Assembly LoadFromPathUnchecked(string fullPath)
         {
+            var assemblyIdentity = AssemblyIdentityUtils.TryGetAssemblyIdentity(fullPath);
+            return LoadFromPathUncheckedCore(fullPath, assemblyIdentity);
+        }
+
+        private Assembly LoadFromPathUncheckedCore(string fullPath, AssemblyIdentity identity)
+        {
             Debug.Assert(PathUtilities.IsAbsolute(fullPath));
 
+            // Check if we have already loaded an assembly with the same identity or from the given path.
+            Assembly loadedAssembly = null;
             lock (_guard)
             {
                 Assembly existingAssembly;
                 if (_loadedAssembliesByPath.TryGetValue(fullPath, out existingAssembly))
                 {
-                    return existingAssembly;
+                    loadedAssembly = existingAssembly;
+                }
+                else if (identity != null && _loadedAssembliesByIdentity.TryGetValue(identity, out existingAssembly))
+                {
+                    loadedAssembly = existingAssembly;
                 }
             }
 
-            Assembly assembly = LoadFromPathImpl(fullPath);
-            return AddToCache(assembly, fullPath);
+            // Otherwise, load the assembly.
+            if (loadedAssembly == null)
+            {
+                loadedAssembly = LoadFromPathImpl(fullPath);
+            }
+
+            // Add the loaded assembly to both path and identity cache.
+            return AddToCache(loadedAssembly, fullPath, identity);
         }
 
-        private Assembly AddToCache(Assembly assembly, string fullPath)
+        private Assembly AddToCache(Assembly assembly, string fullPath, AssemblyIdentity identity)
         {
             Debug.Assert(PathUtilities.IsAbsolute(fullPath));
             Debug.Assert(assembly != null);
 
-            var identity = AssemblyIdentity.FromAssemblyDefinition(assembly);
-
             lock (_guard)
             {
-                // The same assembly may be loaded from two different full paths (e.g. when loaded from GAC, etc.),
-                // or another thread might have loaded the assembly after we checked above.
-                Assembly existingAssembly;
-                if (_loadedAssembliesByIdentity.TryGetValue(identity, out existingAssembly))
+                if (identity != null)
```

Comment:
this seems either identity = identity ?? AssemlbyIdentity.FromAssemblyDefintiion(assembly) or Contract.ThrowIfNull(identity)